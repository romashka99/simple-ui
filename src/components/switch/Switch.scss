@import '../../styles/colors.scss';
@import '../../styles/main.scss';
@import '../../styles/style.scss';

.switch {
	display: flex;
	align-items: center;
	flex-wrap: wrap;
	position: relative;
	cursor: pointer;

	.switch--label {
		padding-left: 8px;

		color: $color-grey;

		font-size: 16px;
		line-height: 24px;
	}

	.switch--inner {
		-webkit-appearance: none;
		box-sizing: border-box;
		outline: none;
		position: relative;
		margin: 0;

		cursor: pointer;

		width: 40px;
		height: 24px;

		background: $color-mid-light-grey;
		border: $border;
		border-radius: 24px;
		border-color: $color-mid-light-grey;

		&:after {
			content: '';
			position: absolute;
			width: 16px;
			height: 16px;
			top: 3px;
			left: 3px;
			background: $color-white;
			border-radius: $border-radius-circle;
		}

		&:checked {
			border-color: $color-grey;
			background: $color-grey;
		}

		&:checked:after {
			left: 19px;
		}

		&:disabled {
			cursor: not-allowed;
			// background: $color-mid-light-grey;
			// border-color: $color-mid-light-grey;

			// &:checked:after {
			// 	background: $color-white;
			// }

			& + .switch--label {
				cursor: not-allowed;
				color: $color-mid-light-grey;
			}
		}

		&:hover {
			&:not(:checked) {
				&:not(:disabled) {
				}
			}
		}
	}

	&.switch--outline {
		& .switch--inner {
			background: $color-white;

			&:after {
				background: $color-mid-light-grey;
			}

			&:checked {
				background: $color-white;
			}

			&:checked:after {
				background: $color-grey;
			}
		}
	}

	&.switch--android {
		& .switch--inner {
			height: 12px;

			&:after {
				top: -7px;
				left: -1px;
				width: 23px;
				height: 23px;
				border: $border;
				border-color: $color-mid-light-grey;
			}

			&:checked:after {
				border-color: $color-grey;
			}

			&:checked:after {
				left: 14px;
			}
		}
	}
}
