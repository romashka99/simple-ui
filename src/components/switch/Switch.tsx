import React, { DetailedHTMLProps, InputHTMLAttributes } from 'react';
import { Color, Size } from '../../types/types';
import './Switch.scss';

interface ISwitchProps
	extends DetailedHTMLProps<
			InputHTMLAttributes<HTMLInputElement>,
			HTMLInputElement
		>,
		React.AriaAttributes {
	color?: Color;
	scale?: Size;
	label?: string;
}

export default function Switch({
	color = 'default',
	scale = 'medium',
	label,
	className,
	...props
}: ISwitchProps) {
	const classes = `switch switch--${color} switch--${scale}${
		className ? ' ' + className : ''
	}`;

	return (
		<>
			<input
				data-testid='switch'
				type='checkbox'
				className={classes}
				{...props}
			/>
			{label && <label>{label}</label>}
		</>
	);
}
