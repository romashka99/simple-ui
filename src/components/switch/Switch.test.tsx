import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Switch from '.';

describe('Test Switch', () => {
	test('without attr', () => {
		render(<Switch />);
		const switcher = screen.getByTestId('switch');
		const inner: HTMLInputElement =
			within(switcher).getByTestId('switch-inner');
		expect(inner).not.toBeNull();
		expect(inner.disabled).toEqual(false);
		expect(switcher.classList.length).toBe(1);
		expect(switcher.classList.contains('switch')).toBe(true);
		expect(switcher).toMatchSnapshot();
	});

	test('has label', () => {
		render(<Switch label='Label' />);
		const switcher = screen.getByTestId('switch');
		const label = within(switcher).getByText('Label');
		expect(label).not.toBeNull();
		expect(switcher).toMatchSnapshot();
	});

	test('checked', () => {
		render(<Switch />);
		const switcher = screen.getByTestId('switch');
		const inner: HTMLInputElement =
			within(switcher).getByTestId('switch-inner');
		expect(inner.checked).toEqual(false);
		userEvent.click(inner);
		expect(inner.checked).toEqual(true);
		expect(switcher).toMatchSnapshot();
	});

	test('disabled', () => {
		render(<Switch disabled />);
		const switcher = screen.getByTestId('switch');
		const inner: HTMLInputElement =
			within(switcher).getByTestId('switch-inner');
		expect(inner.disabled).toEqual(true);
		expect(inner.checked).toEqual(false);
		userEvent.click(inner);
		expect(inner.checked).toEqual(false);
		expect(switcher).toMatchSnapshot();
	});

	test('onChange', () => {
		let value = 0;
		render(
			<Switch
				onChange={() => {
					value = 1;
				}}
			/>
		);
		const switcher = screen.getByTestId('switch');
		expect(value).toBe(0);
		userEvent.click(switcher);
		expect(value).toBe(1);
		expect(switcher).toMatchSnapshot();
	});
});
