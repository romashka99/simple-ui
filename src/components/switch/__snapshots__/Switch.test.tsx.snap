// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Switch without attr 1`] = `
<input
  class="switch switch--default switch--medium"
  data-testid="switch"
  type="checkbox"
/>
`;
